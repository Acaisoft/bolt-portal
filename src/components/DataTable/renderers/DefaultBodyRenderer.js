import React from 'react'
import classNames from 'classnames'

import { Loader } from '~components'
import { TableBody, TableRow, TableCell, Checkbox } from '@material-ui/core'

function DefaultBodyRenderer({
  checkboxes,
  checkboxKey,
  classes,
  columns,
  data,
  handleSelect,
  isLoading,
  rowKey,
  selected,
  striped,
}) {
  if (isLoading) {
    return (
      <TableBody>
        <TableRow className={classNames({ [classes.striped]: striped })}>
          <TableCell colSpan={columns.length}>
            <Loader loading fill />
          </TableCell>
        </TableRow>
      </TableBody>
    )
  }

  return (
    <TableBody>
      {data.map((row, index) => (
        <TableRow
          key={rowKey(row)}
          className={classNames({
            // Indices are zero-based, so we need to shift by one.
            [classes.stripedOdd]: striped && index % 2 === 0,
            [classes.stripedEven]: striped && index % 2 === 1,
          })}
        >
          {checkboxes && (
            <TableCell key="_checkbox" padding="checkbox">
              <Checkbox
                checked={selected && selected.has(checkboxKey(row))}
                onChange={handleSelect(checkboxKey(row))}
              />
            </TableCell>
          )}
          {columns.map(({ render, title, renderFooter, ...cellProps }) => (
            <TableCell {...cellProps}>{render(row)}</TableCell>
          ))}
        </TableRow>
      ))}
    </TableBody>
  )
}

export default DefaultBodyRenderer
