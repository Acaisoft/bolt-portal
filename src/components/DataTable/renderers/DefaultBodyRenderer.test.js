import React from 'react'
import { shallow } from 'enzyme'

import { mocks } from '~utils/tests'

import DefaultBodyRenderer from './DefaultBodyRenderer'
import { children, columnSettings, data } from '../DataTable.test.mock'

const initBodyRenderer = overrides => {
  const mockProps = {
    checkboxKey: row => row.id,
    classes: mocks.ClassesProxy,
    columns: columnSettings,
    data,
    handleSelect: jest.fn(id => jest.fn()),
    rowKey: row => row.id,
    ...overrides,
  }

  const wrapper = shallow(<DefaultBodyRenderer {...mockProps} />)
  return { mockProps, wrapper }
}

describe('component: DefaultBodyRenderer', () => {
  describe('rendering', () => {
    it('should render without crashing', () => {
      const { wrapper } = initBodyRenderer()
      expect(wrapper).toBeTruthy()
    })

    it('should render cells according to column settings', () => {
      const { wrapper } = initBodyRenderer({ children })
      expect(wrapper).toMatchSnapshot()
    })

    it('should render loader if isLoading is set', () => {
      const { wrapper } = initBodyRenderer({ children, isLoading: true })
      expect(wrapper).toMatchSnapshot()
    })

    describe('checkboxes', () => {
      it('should render checkboxes in each row', () => {
        const { wrapper } = initBodyRenderer({ children, checkboxes: true })
        expect(wrapper.find('CheckboxMock').length).toBe(2)
      })

      it('should not render checkboxes if checkboxes prop is set to false', () => {
        const { wrapper } = initBodyRenderer({ children, checkboxes: false })
        expect(wrapper.find('CheckboxMock').length).toBe(0)
      })

      it('should check selected checkboxes', () => {
        const { wrapper } = initBodyRenderer({
          children,
          checkboxes: true,
          selected: new Set([2]),
        })

        expect(wrapper.find('CheckboxMock').at(0)).toHaveProp('checked', false)
        expect(wrapper.find('CheckboxMock').at(1)).toHaveProp('checked', true)
      })

      it('should generate select handler for each checkbox', () => {
        const { mockProps } = initBodyRenderer({
          children,
          checkboxes: true,
        })

        expect(mockProps.handleSelect).toHaveBeenCalledTimes(2)
      })
    })
  })
})
