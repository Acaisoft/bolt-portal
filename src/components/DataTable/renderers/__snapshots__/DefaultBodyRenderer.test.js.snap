// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: DefaultBodyRenderer rendering should render cells according to column settings 1`] = `
<TableBodyMock>
  <TableRowMock
    className=""
    key="1"
  >
    <TableCellMock
      key="name"
    >
      Jane Doe
    </TableCellMock>
    <TableCellMock
      key="age"
    >
      30
    </TableCellMock>
  </TableRowMock>
  <TableRowMock
    className=""
    key="2"
  >
    <TableCellMock
      key="name"
    >
      Steven Bush
    </TableCellMock>
    <TableCellMock
      key="age"
    >
      36
    </TableCellMock>
  </TableRowMock>
</TableBodyMock>
`;

exports[`component: DefaultBodyRenderer rendering should render loader if isLoading is set 1`] = `
<TableBodyMock>
  <TableRowMock
    className=""
  >
    <TableCellMock
      colSpan={2}
    >
      <Loader
        fill={true}
        loading={true}
      />
    </TableCellMock>
  </TableRowMock>
</TableBodyMock>
`;
